### ISSUE TITLE
Maps/map content packs aren't downloaded properly

### ISSUE THREAD

**Issue by hjri** [ceifa/gmod-workshop-downloader#19]
Case in point:
rp_evocity_v5p
https://steamcommunity.com/sharedfiles/filedetails/?id=538207599

it has two content packs + some other materials of questionable sources:
https://steamcommunity.com/workshop/filedetails/?id=538350412
https://steamcommunity.com/workshop/filedetails/?id=538347004

https://steamcommunity.com/sharedfiles/filedetails/?id=2589623653 (not entirely related but seem to have all the materials needed)

All of those are in server's workshop collection, however clients joining do not seem to download the content packs, i had to manually add those ids into autorun lua script - then it worked.

overall it seems that situation with maps is pretty bad, i.e. for metrostroi maps seem to nonchalantly use materials provided by other maps/map-packs, and in case of metrostroi i just used a dumber workshop script that forces client to download literally everything.

At the very least it would be nice if map addons (as map content packs seem to be declared as) would still be sent to client, either unconditionally or if it's in some other sent addon dependency.

**Comment by ceifa** [comment]
Thanks for reporting!
I think the best way to solve this is to check for the required addons and add those as well if subscribed. `steamworks.FileInfo` probably does the job.

**Comment by Xalalau** [comment]
Unfortunately I don't see how to fix the situation with maps depending on maps, it shouldn't even exist in the first place. But I'm going to propose this new map module that should solve the problem with contents:

```lua
local foundCurrentMap = false

function MODULE:Run(context)
    for _, addon in ipairs(context.addons) do
        local hasMapTag = addon.tags and addon.tags:lower():find("map")

        if hasMapTag then
            local bsps = file.Find("maps/*.bsp", addon.title)
            local isContent = #bsps == 0
            local isCurrentMap = false

            -- Add "map content", since we can't check the map it's associated with
            if isContent then
                table.insert(context.usingAddons, addon)
            -- Add the current map
            elseif not foundCurrentMap then
                for _, map in ipairs(bsps) do
                    if map:StripExtension() == currentMap then
                        table.insert(context.usingAddons, addon)
                        foundCurrentMap = true
                        isCurrentMap = true
                        break
                    end
                end
            end
            -- Ignore uloaded maps
            if not isContent and not isCurrentMap then
                context.ignoreResources[addon.wsid] = true
            end
        end
    end
end
```

(Not tested!! Written during a break at work)

**Comment by Xalalau** [comment]
![image](https://user-images.githubusercontent.com/5098527/231951143-e3cdbd40-c85f-4738-9670-c39068db3453.png)

[Add support for module conditions and map contents #25](https://github.com/ceifa/gmod-workshop-downloader/pull/25)

---

Dismissed.

In conclusion, #25:
- Adds unwanted complexibility;
- Works assuming that the Workshop HTML to a degree always stays the same;
- Doesn't work on multiplayer with hybernation turned on. Manually Turning it off may cause unwanted behaviour;
- Has an unpredictable delay in scanning and selecting map content that can lead to some players having to re-enter the game.

**Comment by Xalalau** [comment]
Solved by #26 by letting players manually select map contents.
